lib/spack/spack/new_installer.py

```diff
@@ -0,0 +1,1194 @@
+            with open(pkg.log_path, "rb") as f, open(pkg.install_log_path, "wb") as g:
+                # Use GzipFile directly so we can omit filename / mtime in header
+                gzip_file = GzipFile(filename="", mode="wb", compresslevel=6, mtime=0, fileobj=g)
+                shutil.copyfileobj(f, gzip_file)
+                gzip_file.close()
+
+            spack.hooks.post_install(spec, explicit)
+
+        except Exception:
+            # Print all exceptions so they are logged by the tee thread.
+            traceback.print_exc()
+            exit_code = 1
+        finally:
+            tee.close()
+            state.close()
+
+        sys.exit(exit_code)
+
+
+class JobServer:
+    """Acts both as a jobserver client and server. The server is currently only FIFO-based but
+    can easily be extended to support ordinary pipes."""
+
+    def __init__(self, num_jobs: int) -> None:
+        #: Keep track of how many tokens Spack itself has acquired, which is used to release them.
+        self.tokens_acquired = 0
+        self.num_jobs = num_jobs
+        self.fifo_path: Optional[str] = None
+        self.created = False
+
+        fifo_config = get_jobserver_config()
+
+        if type(fifo_config) is str:
+            # FIFO-based jobserver. Try to open the FIFO.
+            open_attempt = open_existing_jobserver_fifo(fifo_config)
+            if open_attempt:
+                self.r, self.w = open_attempt
+                self.fifo_path = fifo_config
+                return
+        elif type(fifo_config) is tuple:
+            # Old style pipe-based jobserver. Validate the fds before using them.
+            r, w = fifo_config
+            if fcntl.fcntl(r, fcntl.F_GETFD) == -1 or fcntl.fcntl(w, fcntl.F_GETFD) == -1:
+                self.r, self.w = r, w
+                return
+
+        # No existing jobserver we can connect to: create a FIFO-based one.
+        self.r, self.w, self.fifo_path = create_jobserver_fifo(num_jobs)
+        self.created = True
+
+    def acquire(self, jobs: int) -> int:
+        """Try and acquire at most 'jobs' tokens from the jobserver. Returns the number of
+        tokens actually acquired (may be less than requested, or zero)."""
+        try:
+            num_acquired = len(os.read(self.r, jobs))
+            self.tokens_acquired += num_acquired
+            return num_acquired
+        except BlockingIOError:
+            return 0
+
+    def release(self) -> None:
+        """Release a token back to the jobserver."""
+        # The last job to quit has an implicit token, so don't release if we have none.
+        if self.tokens_acquired == 0:
+            return
+        os.write(self.w, b"+")
+        self.tokens_acquired -= 1
+
+    def close(self) -> None:
+        # Remove the FIFO if we created it.
+        if self.created and self.fifo_path:
+            try:
+                os.unlink(self.fifo_path)
+            except OSError:
+                pass
```

Comment:
We should write a level-3 debug message for these as well.

Reply:
done